File: pm.py
from numpy import *
from math import floor
from mpl_toolkits.mplot3d import Axes3D
import matplotlib.pyplot as plt


# function prototypes would go here


def main():

	# cosmological constants
	OmegaM  = 0.27;
	OmegaK  = 0.73;
	OmegaL  = 0.00008;
	H0      = 67.80;   		  #(km/s)/Mpc
	G       = 6.67*(10**-11); # N*m**2 / kg**2

	da = 0.1;  # time-stepping resolution
	a  = 0.01; # start when universe was 0.01 current size
	aMAX = 1.0 # end at current time (when universe is equal to current size)

	# mesh constants
	M = 64**1; # per direction
	L = 1.0;   # actual length of mesh in "suitable" units

	# particle constants
	N  = 32**1;
	mp = 1.0;

	# handy units
	r0 = L / M;    # size of grid cell
	t0 = H0**-1;   # unit of time
	v0 = r0 / t0;  # unit of velocity
	p0 = ( (3*H0**2)/(8*pi*G) ) * OmegaM; # unit of density
	phi0 = v0**2;  # unit of potential

	# Particle arrays
	x = empty(N);
	y = empty(N);
	z = empty(N);

	x_tilde = empty(N);   # don't technically need separate arrays for this
	y_tilde = empty(N);   # tilde arrays are for code units
	z_tilde = empty(N);   # doing this helps clarify the process

	vx = empty(N);
	vy = empty(N);
	vz = empty(N);

	px_tilde = empty(N);
	py_tilde = empty(N);
	pz_tilde = empty(N);

	# Grid arrays
	density       = zeros((M,M,M));
	density_tilde = zeros((M,M,M));
	delta_tilde   = zeros((M,M,M));
	delta_cap     = zeros((M,M,M));   # Fourier space
	Green         = zeros((M,M,M));   # Fourier space
	potential_cap = zeros((M,M,M));   # Fourier space
	potential_tilde = zeros((M,M,M));

	gx_tilde = zeros((M,M,M));   # at cell centers
	gy_tilde = zeros((M,M,M));
	gz_tilde = zeros((M,M,M));

	gx_p = zeros(N);   # acceleration at particle
	gy_p = zeros(N);
	gz_p = zeros(N);

	# assign particles
	for i in range(N):
		x[i] = random.uniform(L)/r0;	# randomly spaced
		y[i] = random.uniform(L)/r0;
		z[i] = random.uniform(L)/r0;

		vx[i] = 0; 				# no initial velocity
		vy[i] = 0;
		vz[i] = 0;

	###########################################################
	###########################################################
	# Plot initial

	fig = plt.figure();       # create figure object
	ax = fig.add_subplot(111, projection='3d'); # create 3D axes

	ax.set_xlim(-10,70);
	ax.set_ylim(-10,70);
	ax.set_zlim(-10,70);

	ax.scatter(x,y,z, c='g');  # plot particles

	ax.set_xlabel('X');
	ax.set_ylabel('Y');
	ax.set_zlabel('Z');

	plt.savefig("0.0.png");
	plt.close();
	############################################################
	############################################################

	t=0;
	TMAX=2;
	da = .01;
	# start while loop (for time)
	while t < TMAX:
		# compute density
		for particle in range(N):
			i = floor(x[particle]);
			j = floor(y[particle]);
			k = floor(z[particle]);

			dx = x[particle] - i;
			dy = y[particle] - j;
			dz = z[particle] - k;

			tx = 1 - dx;
			ty = 1 - dy;
			tz = 1 - dz;
			# Boundary conditions
			ip1 = i+1;
			jp1 = j+1;
			kp1 = k+1;

			if (i == M-1):
				ip1 = 0;
			if (j == M-1):
				jp1 = 0;
			if (k == M-1):
				kp1 = 0;

			density[ i,   j,   k ] += mp*tx*ty*tz;  	# these aren't densities...
			density[ i,  jp1,  k ] += mp*tx*dy*tz;
			density[ i,   j,  kp1] += mp*tx*ty*dz;
			density[ i,  jp1, kp1] += mp*tx*dy*dz;
			density[ip1,  j,   k ] += mp*dx*ty*tz;
			density[ip1, jp1,  k ] += mp*dx*dy*tz;
			density[ip1,  j,  kp1] += mp*dx*ty*dz;
			density[ip1, jp1, kp1] += mp*dx*dy*dz;

		# make density units right
		for i in range(M):
			for j in range(M):
				for k in range(M):
					density_tilde[i,j,k] = density[i,j,k] / p0;

		delta_tilde = density_tilde - 1;

		# Calculate Fourier Transformed density on the mesh
		delta_cap = fft.fftn(delta_tilde);

		# Calculate Green's function
		k = 2*pi / M;
		for l in range(M):
			for m in range(M):
				for n in range(M):
					if (l==0) and (m==0) and (n==0):
						Green[l,m,n] = 0;
					else:
						Green[l,m,n] = -( (3*OmegaM)/(8) )*( sin(.5*k*l)**2 + sin(.5*k*m)**2 + sin(.5*k*n)**2 )**-1;

		# Calculate (FFTed) potential on the mesh using Poisson's eqn
		for l in range(M):
			for m in range(M):
				for n in range(M):
					potential_cap[l,m,n] = Green[l,m,n]*delta_cap[l,m,n];


		# Find potential using an Inverse Fourier Transform
		potential_tilde = fft.ifftn(potential_cap);

		# for all particles do

			# compute field (acceleration) in nearby meshpoints
		for i in range(M):
			for j in range(M):
				for k in range(M):
					ip1 = i+1;  im1 = i-1;
					jp1 = j+1;  jm1 = j-1;
					kp1 = k+1;  km1 = k-1;

					# Boundary setting
					if (i == M-1):
						ip1 = 0;
					if (j == M-1):
						jp1 = 0;
					if (k == M-1):
						kp1 = 0;
					if (i == 0):
						im1 = M-1;
					if (j == 0):
						jm1 = M-1;
					if (k == 0):
						km1 = M-1;

					gx_tilde[i,j,k] = -( potential_tilde[ip1,j,k] - potential_tilde[im1,j,k] )/2;
					gy_tilde[i,j,k] = -( potential_tilde[i,jp1,k] - potential_tilde[i,jm1,k] )/2;
					gz_tilde[i,j,k] = -( potential_tilde[i,j,kp1] - potential_tilde[i,j,km1] )/2;

			# interpolate the field at the particle position
		for particle in range(N):
			i = floor(x[particle]);
			j = floor(y[particle]);
			k = floor(z[particle]);

			dx = x[particle] - i;
			dy = y[particle] - j;
			dz = z[particle] - k;

			tx = 1 - dx;
			ty = 1 - dy;
			tz = 1 - dz;

			# Boundary conditions
			ip1 = i+1;
			jp1 = j+1;
			kp1 = k+1;

			if (i == M-1):
				ip1 = 0;
			if (j == M-1):
				jp1 = 0;
			if (k == M-1):
				kp1 = 0;

			gx_p[particle] =  gx_tilde[i, j,  k ]*tx*ty*tz + gx_tilde[ip1, j,  k ]*dx*ty*tz + \
							  gx_tilde[i,jp1, k ]*tx*dy*tz + gx_tilde[ip1,jp1, k ]*dx*dy*tz + \
							  gx_tilde[i, j, kp1]*tx*ty*dz + gx_tilde[ip1, j, kp1]*dx*ty*dz + \
							  gx_tilde[i,jp1,kp1]*tx*dy*dz + gx_tilde[ip1,jp1,kp1]*dx*dy*dz;

			gy_p[particle] =  gy_tilde[i, j,  k ]*tx*ty*tz + gy_tilde[ip1, j,  k ]*dx*ty*tz + \
							  gy_tilde[i,jp1, k ]*tx*dy*tz + gy_tilde[ip1,jp1, k ]*dx*dy*tz + \
							  gy_tilde[i, j, kp1]*tx*ty*dz + gy_tilde[ip1, j, kp1]*dx*ty*dz + \
							  gy_tilde[i,jp1,kp1]*tx*dy*dz + gy_tilde[ip1,jp1,kp1]*dx*dy*dz;

			gz_p[particle] =  gz_tilde[i, j,  k ]*tx*ty*tz + gz_tilde[ip1, j,  k ]*dx*ty*tz + \
							  gz_tilde[i,jp1, k ]*tx*dy*tz + gz_tilde[ip1,jp1, k ]*dx*dy*tz + \
							  gz_tilde[i, j, kp1]*tx*ty*dz + gz_tilde[ip1, j, kp1]*dx*ty*dz + \
							  gz_tilde[i,jp1,kp1]*tx*dy*dz + gz_tilde[ip1,jp1,kp1]*dx*dy*dz;

			f = 1
			# calculate new speed and position using a suitable scheme
			vx[particle] = vx[particle] + f*gx_p[particle]*da;
			vy[particle] = vy[particle] + f*gy_p[particle]*da;
			vz[particle] = vz[particle] + f*gz_p[particle]*da;

			x[particle] = (x[particle] + f*vx[particle]*da) % 63;
			y[particle] = (y[particle] + f*vy[particle]*da) % 63;
			z[particle] = (z[particle] + f*vz[particle]*da) % 63;

		t = t + da;

		# plotting
		fig = plt.figure()
		ax = fig.add_subplot(111, projection='3d')

		ax.set_xlim(-10,70)
		ax.set_ylim(-10,70)
		ax.set_zlim(-10,70)

		ax.scatter(x,y,z, c='g')  # particles

		ax.set_xlabel('X')
		ax.set_ylabel('Y')
		ax.set_zlabel('Z')

		plt.savefig("{0}.png".format(t))
		plt.close()

	return 0;


def f(a):
	return ( (a**-1) * (OmegaM + OmegaK*a + OmegaL*a**3) )**(-0.5);



if __name__ == __main__:
	main()

File: pm.cxx
///////////////////////////////////////////////////////////////////////////////
/**  pm.cxx ::  PM N-body code
 *   
 *   Authors: Joseph, Chunyang, Mehmet
 *	 Date   : May 9. 2016
 *   *********************************
 *	
 *   This code models the gravitational interaction of many bodies in a
 *   collisionless system via the Particle-Mesh method.
 *
 *   FILES:
 *
 *   	pm.cxx is the main function.
 *   	pm.h   contains global variables and function prototypes
 *
 *   	data is written out to 'particle_positions.txt' and 'velocity.txt'
 *
 *   	This data is read by 'plotting.py' and 'plot_vx_vs_v.py' respectively
 *
 *   	These files generate plots of the data at each timestep which are 
 *   	saved in the directories 3D and VX.
 *
 *	 KNOWN-ISSUES:
 *
 *		D+ :: We don't have a clue what this is, but we have it set to 5
 *			  (in general the whole initial condition is a little shakey)
 *		
 *		Leap-frog Integration :: another thing we don't have a clue about
 *            (we believe we are currently using 1st-order Euler scheme)
 *
 *
 *
 */
///////////////////////////////////////////////////////////////////////////////
#include <iostream>
#include <fstream>
#include <cmath>    // pow(), sin()
#include <fftw3.h>
#include <cstdlib>  // srand48(), drand48()

#include "pm.h"

using namespace std;

int main()
{


// time-stepping variables
	  double a    = 0.010;
const double aMAX = 0.200;
const double da   = 0.001;

// particle arrays :: N = nparticle
double x[N];
double y[N];
double z[N];

double vx[N];
double vy[N];
double vz[N];

// grid arrays (fftw uses its own malloc())
double *rho;
double *phi;

rho = (double*) fftw_malloc(sizeof(double)*Mtot);
phi = (double*) fftw_malloc(sizeof(double)*Mtot);




///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////



/* setup initial conditions */


double Dplusdot=1; // basically useless, but don't change it
Dplus = 5;

/** 
 *  Particle locations are distributed randomly through out the 64x64x64 grid
 *  Then they are shifted in the Zel'dovich approximation. Only the x-component
 *  is shifted (1D-plane wave). We aren't totally confident about this.
 */


// set seed
srand48(1);

// loop over particles
for (int i=0; i<N; i++){
	x[i]=drand48()*M1D;
	y[i]=drand48()*M1D;
	z[i]=drand48()*M1D;



	x[i]= ((i-1)*M1D / N) + Dplus * sin(2*pi*((i-1)*M1D / N) / M1D);


	vx[i]=Dplusdot*sin(2*pi*x[i]/M1D);
	vy[i]=0;
	vz[i]=0;
}



///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////


/* set-up output files */

ofstream VXoutFile, out3D;
out3D.open("particle_positions.txt");
VXoutFile.open("velocity.txt");

/***********************/

// main

/* loop over time steps */
while ( a < aMAX )
{
	cicInterpolate(x,y,z, rho);

	solvePoisson(a, rho, phi);

	// print out data
	for (int i=0; i<N; i++){
		out3D   << x[i] << ' '
				<< y[i] << ' '
				<< z[i] << ' '
						<< '\n';
	}

	for (int i=0; i<N; i++){
		VXoutFile << x[i]  << ' '
				  << vx[i] << ' '
						   << '\n';
	}

	updateParticles(a,da, x,y,z, vx,vy,vz, phi);


	a += da;
}



/* close output files */
VXoutFile.close();
out3D.close();


/* free memory */
fftw_free(rho);
fftw_free(phi);


	return 0;
}










void cicInterpolate(double *x, double *y, double *z, double *rho){

  int I, J, K;      // Parent cell coordinates

  double dx,dy,dz;  // distances from parent cell center to particle

  double tx,ty,tz;  // interpolation factors


  /* loop over particles and apply the CIC interpolation to calculate density of each cell */
  for (int particle=0; particle<N; particle++)
  {
    I = floor(x[particle]);
    J = floor(y[particle]);
    K = floor(z[particle]);

    dx = x[particle] - (double) I;    tx = 1 - dx;
    dy = y[particle] - (double) J;    ty = 1 - dy;
    dz = z[particle] - (double) K;    tz = 1 - dz;


    rho[   (I %M1D)   *M1D*M1D +   (J %M1D)   *M1D +   (K %M1D)]   += tx*ty*tz;
    rho[   (I %M1D)   *M1D*M1D + ((J+1) %M1D) *M1D +   (K %M1D)]   += tx*dy*tz;
    rho[   (I %M1D)   *M1D*M1D +   (J %M1D)   *M1D + ((K+1) %M1D)] += tx*ty*dz;
    rho[   (I %M1D)   *M1D*M1D + ((J+1) %M1D) *M1D + ((K+1) %M1D)] += tx*dy*dz;
    rho[ ((I+1) %M1D) *M1D*M1D +   (J %M1D)   *M1D +   (K %M1D)]   += dx*ty*tz;
    rho[ ((I+1) %M1D) *M1D*M1D + ((J+1) %M1D) *M1D +   (K %M1D)]   += dx*dy*tz;
    rho[ ((I+1) %M1D) *M1D*M1D +   (J %M1D)   *M1D + ((K+1) %M1D)] += dx*ty*dz;
    rho[ ((I+1) %M1D) *M1D*M1D + ((J+1) %M1D) *M1D + ((K+1) %M1D)] += dx*dy*dz;
  }

}


void solvePoisson(double a, double *rho, double *phi) {
/**
 *   I am just going to say it. This function is a mess.
 *   I am 95% sure it is corrent.
 *
 *   The order that things are declared and executed is really important
 *   and the size of the arrays was a pain and a half to find out.
 *
 */





	fftw_complex *frho;
	const int frho_size = M1D*M1D*(M1D/2 + 1);
	frho = (fftw_complex*) fftw_malloc(sizeof(fftw_complex)*frho_size);

	/* fourier transform rho into fourier space */

	// create plan (let fftw3 learn machine architecture)
	fftw_plan pf;
	pf = fftw_plan_dft_r2c_3d(M1D, M1D, M1D, rho, frho, FFTW_ESTIMATE);

	fftw_execute(pf);
	fftw_destroy_plan(pf);



	/* calculate green's function in fourier space */
	double G[frho_size];
	double kx,ky,kz;

	for (int l=0; l<M1D; l++){
	for (int m=0; m<M1D; m++){
	for (int n=0; n<(M1D/2 + 1); n++){

		// frequencies
		kx = 2*pi*(l - (M1D/2)) / M1D;
		ky = 2*pi*(m - (M1D/2)) / M1D;
		kz = 2*pi*n / M1D;

		// done out here to save horizontal line space
		double sin2kx = pow(sin(kx/2.0), 2);
		double sin2ky = pow(sin(ky/2.0), 2);
		double sin2kz = pow(sin(kz/2.0), 2);

		// set (0,0,0) component to 0
		if ( (kx==0) && (ky==0) && (kz==0) ) {
			G[l*M1D*(M1D/2 + 1) + m*(M1D/2 + 1) + n] = 0;
		}
		else {
			G[l*M1D*(M1D/2 + 1) + m*(M1D/2 + 1) + n] = -((3.0*OmegaM)/(8.0*a))*pow( (sin2kx + sin2ky + sin2kz) ,-1);
		}
	}}}


	// get phi array ready

	fftw_complex *fphi;
	fphi = (fftw_complex*) fftw_malloc(sizeof(fftw_complex)*frho_size);

	// make plan for backward transformation
	fftw_plan pb;
	pb = fftw_plan_dft_c2r_3d(M1D, M1D, M1D, fphi, phi, FFTW_ESTIMATE);

	// calculated necessary stuff
	for (int l=0; l<M1D; l++){
	for (int m=0; m<M1D; m++){
	for (int n=0; n<(M1D/2 + 1); n++){
			int pos=l*M1D*(M1D/2 + 1) + m*(M1D/2 + 1) + n;
		
		fphi[pos][0] = G[pos]*frho[pos][0];
		fphi[pos][1] = G[pos]*frho[pos][1];
			}
		}	
	}



	/* reverse transformation using fftw_plan_dft_c2r_3d() and fftw_execute() */

	fftw_execute(pb);
	fftw_destroy_plan(pb);

	/* normalize phi */
	for (int i=0; i<Mtot; i++){
		phi[i]= (1/(double)Mtot)*phi[i];
	}


	/* free memory */
	fftw_free(frho);
	fftw_free(fphi);
}



void updateParticles(double a, double da, double *x, double *y, double *z, double *vx, double *vy, double *vz, double *phi)
{
	/* declarations of relevant variables*/
	int ip1,im1,jp1,jm1,kp1,km1;

	double *gx;
	double *gy;
	double *gz;

	// potential ON THE MESH
	gx = (double*) malloc(sizeof(double)*Mtot);
	gy = (double*) malloc(sizeof(double)*Mtot);
	gz = (double*) malloc(sizeof(double)*Mtot);




	int I, J, K;      // Parent cell coordinates
	int IP1, JP1, KP1;

	double dx,dy,dz;  // distances from parent cell center to particle

	double tx,ty,tz;  // interpolation factors

	// accelerations of particle
	double ax;
	double ay;
	double az;


	// calculate gradient of potential ON THE MESH
	for (int i=0; i<M1D; i++){
	for (int j=0; j<M1D; j++){
	for (int k=0; k<M1D; k++){
		ip1 = i+1;  im1 = i-1;
		jp1 = j+1;  jm1 = j-1;
		kp1 = k+1;  km1 = k-1;

		// Boundary setting
		if (i == M1D-1) ip1 = 0;
		if (j == M1D-1) jp1 = 0;
		if (k == M1D-1) kp1 = 0;
		if (i == 0) im1 = M1D-1;
		if (j == 0) jm1 = M1D-1;
		if (k == 0) km1 = M1D-1;

		gx[i*M1D*M1D + j*M1D +k]= -.5*(phi[ip1*M1D*M1D +  j*M1D  +  k ] - phi[im1*M1D*M1D + j*M1D  + k ]);
		gy[i*M1D*M1D + j*M1D +k]= -.5*(phi[ i*M1D*M1D  + jp1*M1D +  k ] - phi[ i*M1D*M1D + jm1*M1D + k ]);
		gz[i*M1D*M1D + j*M1D +k]= -.5*(phi[ i*M1D*M1D  +  j*M1D  + kp1] - phi[ i*M1D*M1D  + j*M1D + km1]); 

	}}}


	/* calculate particle accelerations from grad phi */
	for (int p=0; p<N; p++){

		I = floor(x[p]);    IP1 = I+1;
		J = floor(y[p]);	JP1 = J+1;
		K = floor(z[p]);	KP1 = K+1;

		if (IP1 == M1D) IP1 = 0;
		if (JP1 == M1D) JP1 = 0;
		if (KP1 == M1D) KP1 = 0;

		dx = x[p] - (double) I;    tx = 1 - dx;
		dy = y[p] - (double) J;    ty = 1 - dy;
		dz = z[p] - (double) K;    tz = 1 - dz;

		ax = gx[I*M1D*M1D + J*M1D + K]*tx*ty*tz + gx[IP1*M1D*M1D + J*M1D + K]*dx*ty*tz +
			 gx[I*M1D*M1D + JP1*M1D + K]*tx*dy*tz + gx[IP1*M1D*M1D + JP1*M1D + K]*dx*dy*tz +
			 gx[I*M1D*M1D + J*M1D + KP1]*tx*ty*dz + gx[IP1*M1D*M1D + J*M1D + KP1]*dx*ty*dz +
			 gx[I*M1D*M1D + JP1*M1D + KP1]*tx*dy*dz + gx[IP1*M1D*M1D + JP1*M1D + KP1]*dx*dy*dz;

		ay = gy[I*M1D*M1D + J*M1D + K]*tx*ty*tz + gy[IP1*M1D*M1D + J*M1D + K]*dx*ty*tz +
			 gy[I*M1D*M1D + JP1*M1D + K]*tx*dy*tz + gy[IP1*M1D*M1D + JP1*M1D + K]*dx*dy*tz +
			 gy[I*M1D*M1D + J*M1D + KP1]*tx*ty*dz + gy[IP1*M1D*M1D + J*M1D + KP1]*dx*ty*dz +
			 gy[I*M1D*M1D + JP1*M1D + KP1]*tx*dy*dz + gy[IP1*M1D*M1D + JP1*M1D + KP1]*dx*dy*dz;

		az = gz[I*M1D*M1D + J*M1D + K]*tx*ty*tz + gz[IP1*M1D*M1D + J*M1D + K]*dx*ty*tz +
			 gz[I*M1D*M1D + JP1*M1D + K]*tx*dy*tz + gz[IP1*M1D*M1D + JP1*M1D + K]*dx*dy*tz +
			 gz[I*M1D*M1D + J*M1D + KP1]*tx*ty*dz + gz[IP1*M1D*M1D + J*M1D + KP1]*dx*ty*dz +
			 gz[I*M1D*M1D + JP1*M1D + KP1]*tx*dy*dz + gz[IP1*M1D*M1D + JP1*M1D + KP1]*dx*dy*dz;


		/* update particle velocities */
		vx[p] += f(a) * ax * da;
		vy[p] += f(a) * ay * da;
		vz[p] += f(a) * az * da;
		/* update particle positions */ 
		x[p] += pow((a+da/2.0), -2) * f(a + da/2.0) * vx[p] * da;
		y[p] += pow((a+da/2.0), -2) * f(a + da/2.0) * vy[p] * da;
		z[p] += pow((a+da/2.0), -2) * f(a + da/2.0) * vz[p] * da;

	}


	free(gx);
	free(gy);
	free(gz);


}

// helper function which prints out 3D arrays
void printVec3D(double* rho) {
  for (int i=0; i<M1D; i++) {
    for (int j=0; j<M1D; j++) {
      for (int k=0; k<M1D; k++) {
        cout << rho[i*M1D*M1D + j*M1D + k] << ' ';
        }
      cout << endl;
    }
  cout << endl;
  }
}

double f(double a){
  return pow((1 / a) * sqrt( OmegaM + OmegaK * a + OmegaL * pow(a, 2) ), -0.5);
}

